From 43ec8a89f6c99130727084b3496a2ffd2b4200f0 Mon Sep 17 00:00:00 2001
From: Karel Zak <kzak@redhat.com>
Date: Mon, 31 Aug 2026 17:03:32 +0200
Subject: [PATCH] nsenter, unshare: add O_CLOEXEC to all open() calls
 [CVE-2026-78408]

Add O_CLOEXEC (and EFD_CLOEXEC for eventfd) as defense in depth to
all file descriptor creation sites in nsenter and unshare. All these
descriptors are already explicitly closed before exec, but O_CLOEXEC
provides a safety net against future code changes that might
accidentally introduce a leak path.

No functional change.

Signed-off-by: Karel Zak <kzak@redhat.com>
---
 sys-utils/nsenter.c |  8 ++++----
 sys-utils/unshare.c | 12 ++++++------
 2 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/sys-utils/nsenter.c b/sys-utils/nsenter.c
index cf6c831744b..9d9d90a4809 100644
--- a/sys-utils/nsenter.c
+++ b/sys-utils/nsenter.c
@@ -171,7 +171,7 @@ static void open_target_fd(int *fd, const char *type, const char *path)
 	if (*fd >= 0)
 		close(*fd);
 
-	*fd = open(path, O_RDONLY);
+	*fd = open(path, O_RDONLY | O_CLOEXEC);
 	if (*fd < 0)
 		err(EXIT_FAILURE, _("cannot open %s"), path);
 }
@@ -396,7 +396,7 @@ static void open_cgroup_procs(void)
 
 	snprintf(fdpath, sizeof(fdpath), _PATH_SYS_CGROUP "/%s/cgroup.procs", path);
 
-	if ((cgroup_procs_fd = open(fdpath, O_WRONLY | O_APPEND)) < 0)
+	if ((cgroup_procs_fd = open(fdpath, O_WRONLY | O_APPEND | O_CLOEXEC)) < 0)
 		err(EXIT_FAILURE, _("failed to open cgroup.procs"));
 
 	free(buf);
@@ -762,7 +762,7 @@ int main(int argc, char *argv[])
 
 	/* Remember the current working directory if I'm not changing it */
 	if (root_fd >= 0 && wd_fd < 0 && wdns == NULL) {
-		wd_fd = open(".", O_RDONLY);
+		wd_fd = open(".", O_RDONLY | O_CLOEXEC);
 		if (wd_fd < 0)
 			err(EXIT_FAILURE,
 			    _("cannot open current working directory"));
@@ -785,7 +785,7 @@ int main(int argc, char *argv[])
 
 	/* working directory specified as in-namespace path */
 	if (wdns) {
-		wd_fd = open(wdns, O_RDONLY);
+		wd_fd = open(wdns, O_RDONLY | O_CLOEXEC);
 		if (wd_fd < 0)
 			err(EXIT_FAILURE,
 			    _("cannot open current working directory"));
diff --git a/sys-utils/unshare.c b/sys-utils/unshare.c
index 05db627bef5..c477398625e 100644
--- a/sys-utils/unshare.c
+++ b/sys-utils/unshare.c
@@ -108,7 +108,7 @@ static void setgroups_control(int action)
 		return;
 	cmd = setgroups_strings[action];
 
-	fd = open(file, O_WRONLY);
+	fd = open(file, O_WRONLY | O_CLOEXEC);
 	if (fd < 0) {
 		if (errno == ENOENT)
 			return;
@@ -125,7 +125,7 @@ static void map_id(const char *file, uint32_t from, uint32_t to)
 	char *buf;
 	int fd;
 
-	fd = open(file, O_WRONLY);
+	fd = open(file, O_WRONLY | O_CLOEXEC);
 	if (fd < 0)
 		 err(EXIT_FAILURE, _("cannot open %s"), file);
 
@@ -219,7 +219,7 @@ static void settime(int64_t offset, clockid_t clk_id)
 
 	len = snprintf(buf, sizeof(buf), "%d %" PRId64 " 0", clk_id, offset);
 
-	fd = open("/proc/self/timens_offsets", O_WRONLY);
+	fd = open("/proc/self/timens_offsets", O_WRONLY | O_CLOEXEC);
 	if (fd < 0)
 		err(EXIT_FAILURE, _("failed to open /proc/self/timens_offsets"));
 
@@ -289,7 +289,7 @@ static pid_t fork_and_wait(int *fd)
 	pid_t pid;
 	uint64_t ch;
 
-	*fd = eventfd(0, 0);
+	*fd = eventfd(0, EFD_CLOEXEC);
 	if (*fd < 0)
 		err(EXIT_FAILURE, _("eventfd failed"));
 
@@ -744,11 +744,11 @@ static void load_interp(const char *binfmt_mnt, const char *interp)
 {
 	int dirfd, fd;
 
-	dirfd = open(binfmt_mnt, O_PATH | O_DIRECTORY);
+	dirfd = open(binfmt_mnt, O_PATH | O_DIRECTORY | O_CLOEXEC);
 	if (dirfd < 0)
 		err(EXIT_FAILURE, _("cannot open %s"), binfmt_mnt);
 
-	fd = openat(dirfd, "register", O_WRONLY);
+	fd = openat(dirfd, "register", O_WRONLY | O_CLOEXEC);
 	if (fd < 0)
 		err(EXIT_FAILURE, _("cannot open %s/register"), binfmt_mnt);
 
